### PR TITLE
feat: subsidy access policy models

### PR DESCRIPTION
**Pull request by muhammad-ammar**
**Description:** This refactored the existing models as per the new [proxy model design specification](https://2u-internal.atlassian.net/wiki/spaces/SOL/pages/337805331/Ideas+Policy+django+models+and+API+routes.).

### CODE REVIEW

File: enterprise_access/apps/subsidy_access_policy/models.py [hunk 1]
```diff
@@ -11,18 +11,21 @@
...
     .. no_pii:
     """
-    class Meta:
-        """
-        Metaclass for SubsidyAccessPolicy.
-        """
-        abstract = True
+
+    PROXY_FIELD_NAME = 'proxy_class_name'
+    proxy_class_name = models.CharField(max_length=64)
```
Comment (irfanuddinahmad) [inline enterprise_access/apps/subsidy_access_policy/models.py:28]: the name is too django centric ... can we call this policy_name or something similar?

File: enterprise_access/apps/subsidy_access_policy/models.py [hunk 2]
```diff
@@ -179,23 +193,23 @@ def can_redeem(self, learner_id, content_key):
...
+class CappedEnrollmentLearnerCreditAccessPolicy(SubsidyAccessPolicy):
     """
     Policy that limits the maximum amount that can be spent aggregated across all users covered by this policy
     """
-    spend_limit = models.IntegerField(
-        blank=True,
-        default=0,
-    )
+
+    objects = PolicyManager()
```
Comment (irfanuddinahmad) [inline enterprise_access/apps/subsidy_access_policy/models.py:201]: should we name this after the class?

Comment (muhammad-ammar) [reply enterprise_access/apps/subsidy_access_policy/models.py:201]: This manager will be used by all the different sub-classes. Should we name it after parent class?

File: enterprise_access/apps/subsidy_access_policy/models.py [hunk 3]
```diff
@@ -11,45 +11,83 @@
...
     .. no_pii:
     """
-    class Meta:
-        """
-        Metaclass for SubsidyAccessPolicy.
-        """
-        abstract = True
+
+    POLICY_FIELD_NAME = 'policy_name'
+    policy_name = models.CharField(max_length=64)
```
Comment (sameenfatima78) [inline enterprise_access/apps/subsidy_access_policy/models.py:28]: nit: shouldn't it be called `policy_type` according to the diagram here https://2u-internal.atlassian.net/wiki/spaces/SOL/pages/300580900/EMET+Scrum+of+Scrums+Notes

**Review by sameenfatima78 — APPROVED** [review]
looks good, just one comment

File: enterprise_access/apps/subsidy_access_policy/models.py [hunk 4]
```diff
@@ -11,45 +11,88 @@
...
-    def enterprise_customer_uuid(self):
-        "Returns the enterprise_customer_uuid. Equivalent to the group uuid at present"
-        return self.group_uuid
+    def save(self, *args, **kwargs):
+        """
+        Override to persist policy type.
+        """
+        # TODO: find a better way to do this. Raising an exception will cause 500
+        if type(self).__name__ == SubsidyAccessPolicy.__name__:
+            # it doesn't make sense to create an object of SubsidyAccessPolicy2 because it is not a concrete policy
```
Comment (irfanuddinahmad) [inline enterprise_access/apps/subsidy_access_policy/models.py:71]: remove the '2' postfix

File: enterprise_access/apps/subsidy_access_policy/models.py [hunk 5]
```diff
@@ -11,45 +11,88 @@
...
+            raise TypeError("Can not create object of class SubsidyAccessPolicy2")
```
Comment (irfanuddinahmad) [inline enterprise_access/apps/subsidy_access_policy/models.py:72]: same as above

**Review by irfanuddinahmad — APPROVED** [review]

**Pull request merged**